File: packages/@axonpack/expo-devtools/ios/AxonpackDevtoolsModule.swift
import ExpoModulesCore
import UIKit

/**
 Frames the *main* thread actually rendered. A `requestAnimationFrame` loop in JS cannot see this: the UI
 thread can be fully stalled while JS keeps ticking happily, which is the blind spot the FPS card warns
 about. `CADisplayLink` fires on the main runloop, so it stops counting exactly when the UI stops drawing.
 */
private final class UiFpsTracker: NSObject {
  private var displayLink: CADisplayLink?
  private var frames = 0
  private var windowStart = CACurrentMediaTime()
  private(set) var fps: Double = -1

  func start() {
    guard displayLink == nil else { return }
    frames = 0
    windowStart = CACurrentMediaTime()
    let link = CADisplayLink(target: self, selector: #selector(tick))
    link.add(to: .main, forMode: .common)
    displayLink = link
  }

  func stop() {
    displayLink?.invalidate()
    displayLink = nil
    fps = -1
  }

  @objc private func tick() {
    frames += 1
    let now = CACurrentMediaTime()
    let elapsed = now - windowStart
    if elapsed >= 0.5 {
      fps = Double(frames) / elapsed
      frames = 0
      windowStart = now
    }
  }
}

/**
 The only native code in this package, and it exists to *cause* problems rather than measure them:
 the Limiter section needs to stall or kill the main thread, which JavaScript cannot reach.

 Everything here is optional on the JS side (`requireOptionalNativeModule`), so an app without a dev
 client keeps working — it just can't reach the main thread.
 */
/**
 Read once via `sysctl`, which is the only way to learn when the process actually began — every clock
 reachable from JS starts long after that, which is why `performance.rnStartupTiming` can only report
 what the platform chose to record.
 */
private func processStartEpochMs() -> Double? {
  var info = kinfo_proc()
  var size = MemoryLayout<kinfo_proc>.stride
  var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
  guard sysctl(&mib, u_int(mib.count), &info, &size, nil, 0) == 0 else {
    return nil
  }
  let started = info.kp_proc.p_starttime
  return Double(started.tv_sec) * 1000 + Double(started.tv_usec) / 1000
}

/**
 The app's real memory footprint — what iOS actually holds against you, and what a user means by
 "memory". `phys_footprint` is the same number Xcode's memory gauge shows, and it is unrelated to the JS
 heap the Performance tab reads from Hermes.
 */
private func appMemoryFootprintBytes() -> Double? {
  var info = task_vm_info_data_t()
  var count = mach_msg_type_number_t(MemoryLayout<task_vm_info_data_t>.size) / 4
  let result = withUnsafeMutablePointer(to: &info) {
    $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
      task_info(mach_task_self_, task_flavor_t(TASK_VM_INFO), $0, &count)
    }
  }
  guard result == KERN_SUCCESS else { return nil }
  return Double(info.phys_footprint)
}


/**
 Crash persistence.

 An uncaught Objective-C exception takes the process down before JavaScript gets another turn, so a
 report has to be written from the dying thread — this is the one thing in crash reporting that JS
 cannot do for itself. Written as JSON Lines and appended: a crash handler is the wrong place to be
 reading a file back in to rewrite it.

 Deliberately **not** a POSIX signal handler. `NSSetUncaughtExceptionHandler` runs in ordinary
 context, so normal Foundation calls are safe here; a `SIGSEGV` handler would have to be
 async-signal-safe and would fight Crashlytics/Sentry over the same slot. Swift `fatalError` and
 memory faults are therefore out of scope, and the package says so rather than implying coverage.
 */
private let crashFileName = "axonpack-devtools-crashes.jsonl"

private func crashFileURL() -> URL? {
  let manager = FileManager.default
  guard
    let directory = manager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
  else {
    return nil
  }
  // Application Support is not created for you, and unlike Caches it is not purged under disk
  // pressure — a crash report that evaporates before the next launch is worse than none.
  if !manager.fileExists(atPath: directory.path) {
    try? manager.createDirectory(at: directory, withIntermediateDirectories: true)
  }
  return directory.appendingPathComponent(crashFileName)
}

private func appendCrashLine(_ json: String) {
  guard let url = crashFileURL(), let data = (json + "\n").data(using: .utf8) else { return }

  if let handle = try? FileHandle(forWritingTo: url) {
    defer { try? handle.close() }
    _ = try? handle.seekToEnd()
    try? handle.write(contentsOf: data)
  } else {
    try? data.write(to: url)
  }
}

private func jsonEscape(_ value: String) -> String {
  var escaped = ""
  for character in value.unicodeScalars {
    switch character {
    case "\"": escaped += "\\\""
    case "\\": escaped += "\\\\"
    case "\n": escaped += "\\n"
    case "\r": escaped += "\\r"
    case "\t": escaped += "\\t"
    default:
      if character.value < 0x20 {
        escaped += String(format: "\\u%04x", character.value)
      } else {
        escaped.unicodeScalars.append(character)
      }
    }
  }
  return escaped
}

/// The handler that was installed before ours, called after we persist so Crashlytics/Sentry — or
/// React Native's own — still see the exception.
private var previousExceptionHandler: (@convention(c) (NSException) -> Void)?

private func writeExceptionRecord(_ exception: NSException) {
  let frames = exception.callStackSymbols.map { "\"\(jsonEscape($0))\"" }.joined(separator: ",")
  let json = """
    {"kind":"native-exception",\
    "name":"\(jsonEscape(exception.name.rawValue))",\
    "message":"\(jsonEscape(exception.reason ?? ""))",\
    "stack":null,\
    "timestamp":\(Int(Date().timeIntervalSince1970 * 1000)),\
    "native":{"type":"\(jsonEscape(exception.name.rawValue))","thread":"\(Thread.isMainThread ? "main" : "background")","frames":[\(frames)]}}
    """
  appendCrashLine(json)
}

/// A C function pointer, so it must capture nothing — everything it needs is in the file-scope
/// globals above.
private let axonpackExceptionHandler: @convention(c) (NSException) -> Void = { exception in
  writeExceptionRecord(exception)
  previousExceptionHandler?(exception)
}

private func deviceModelIdentifier() -> String {
  var systemInfo = utsname()
  uname(&systemInfo)
  let mirror = Mirror(reflecting: systemInfo.machine)
  return mirror.children.reduce(into: "") { identifier, element in
    guard let value = element.value as? Int8, value != 0 else { return }
    identifier += String(UnicodeScalar(UInt8(value)))
  }
}


/**
 Connection-phase timing.

 The phases inside a request — queueing, DNS, TCP, TLS — are measured by URLSession and reported
 nowhere JavaScript can reach. React Native's own `PerformanceResourceTiming` looks like the JS
 answer and is not: it fills those fields from the same three instants a JS patch already sees, so it
 reports queueing as zero and connection setup as the whole wait. `URLSessionTaskMetrics` is the real
 measurement, and it arrives through an *optional* delegate method that the session delegates in an
 app do not implement.

 So the method is **added** to those delegate classes at runtime rather than swizzled over them.
 Adding is strictly observational: nothing existing is replaced, no request can be altered or delayed,
 and a class that already implements the method is left alone rather than fought over — another tool
 collecting the same metrics keeps working, and this one reports nothing for that stack.
 */
private func millisBetween(_ start: Date?, _ end: Date?) -> Double? {
  guard let start, let end else { return nil }
  let elapsed = end.timeIntervalSince(start) * 1000
  return elapsed >= 0 ? elapsed : nil
}

private final class NetworkPhaseReporter {
  static let shared = NetworkPhaseReporter()

  /// Set while the reporter is installed, and the only thing that decides whether metrics are sent.
  var emit: (([String: Any?]) -> Void)?
  /// Whether the delegate method is in place. Adding it twice fails by design, so this remembers.
  private(set) var installed = false

  func markInstalled() {
    installed = true
  }

  func record(_ metrics: URLSessionTaskMetrics) {
    guard let emit else { return }
    // The last transaction is the one that produced the response; any earlier ones are redirects,
    // each of which the app saw as part of the same request.
    guard
      let transaction = metrics.transactionMetrics.last,
      let url = transaction.request.url?.absoluteString,
      let fetchStart = transaction.fetchStartDate
    else {
      return
    }

    // Whichever of these the stack reached first is where the request stopped waiting to be worked
    // on, so the gap before it is the queue.
    let workStart =
      transaction.domainLookupStartDate ?? transaction.connectStartDate
      ?? transaction.requestStartDate

    emit([
      "url": url,
      "startMs": fetchStart.timeIntervalSince1970 * 1000,
      "queuedMs": millisBetween(fetchStart, workStart),
      "dnsMs": millisBetween(transaction.domainLookupStartDate, transaction.domainLookupEndDate),
      // TLS is measured separately below, so the connect phase stops where the handshake starts.
      "tcpMs": millisBetween(
        transaction.connectStartDate,
        transaction.secureConnectionStartDate ?? transaction.connectEndDate),
      "tlsMs": millisBetween(
        transaction.secureConnectionStartDate, transaction.secureConnectionEndDate),
      "sendMs": millisBetween(transaction.requestStartDate, transaction.requestEndDate),
      "waitMs": millisBetween(
        transaction.requestEndDate ?? transaction.requestStartDate, transaction.responseStartDate),
      "downloadMs": millisBetween(transaction.responseStartDate, transaction.responseEndDate),
      "reusedConnection": transaction.isReusedConnection,
      "protocol": transaction.networkProtocolName,
      "measuredBy": "urlsession",
    ])
  }
}

private let metricsSelector = NSSelectorFromString("URLSession:task:didFinishCollectingMetrics:")

/**
 The *session* delegate classes worth asking, which is the distinction that matters: URLSession reports
 metrics to the delegate its session was created with, never to a per-task delegate that one forwards
 to. React Native's is Objective-C and public, so it is reliable. Expo's fetch runs through a proxy in
 `ExpoModulesCore`, reached by its Swift runtime name and simply missed if Expo renames it — both
 spellings are tried, because a Swift class registers under a mangled symbol while the `Module.Class`
 form is resolved by the runtime instead.

 `Expo.ExpoURLSessionTask` was the first attempt and was wrong: it is the per-task delegate the proxy
 forwards to, so it resolved, accepted the method, and was never called.
 */
private let sessionDelegateClassNames = [
  "RCTHTTPRequestHandler",
  "ExpoModulesCore.URLSessionSessionDelegateProxy",
  "_TtC15ExpoModulesCore30URLSessionSessionDelegateProxy",
]

private func addMetricsCollection(to className: String) -> Bool {
  guard let delegateClass = NSClassFromString(className) else { return false }
  // Already there: either a future React Native implements it, or another tool got here first.
  if class_getInstanceMethod(delegateClass, metricsSelector) != nil { return false }

  let collect:
    @convention(block) (AnyObject, URLSession, URLSessionTask, URLSessionTaskMetrics) -> Void = {
      _, _, _, metrics in
      NetworkPhaseReporter.shared.record(metrics)
    }

  // "v@:@@@" — returns void, takes self, the selector, and the three objects the delegate method is
  // handed. URLSession asks `respondsToSelector:` before calling, so adding the method is all it takes.
  return class_addMethod(
    delegateClass, metricsSelector, imp_implementationWithBlock(collect), "v@:@@@")
}

public class AxonpackDevtoolsModule: Module {
  /// Captured when the module is constructed, which happens during native startup.
  private let moduleInitEpochMs = Date().timeIntervalSince1970 * 1000

  private let uiFps = UiFpsTracker()

  public func definition() -> ModuleDefinition {
    Name("AxonpackDevtools")

    Events("onNetworkPhases")

    /**
     Returns whether phases will actually be reported, which is what the Timing tab says out loud.
     False means no delegate class could be reached — an app whose networking this build cannot see —
     and the tab keeps to the two numbers a patch measures rather than implying more.
     */
    Function("installNetworkTimingReporter") { () -> Bool in
      NetworkPhaseReporter.shared.emit = { [weak self] payload in
        self?.sendEvent("onNetworkPhases", payload)
      }
      if NetworkPhaseReporter.shared.installed { return true }

      let hooked = sessionDelegateClassNames.filter { addMetricsCollection(to: $0) }
      if hooked.isEmpty { return false }
      NetworkPhaseReporter.shared.markInstalled()
      return true
    }

    // Started and stopped by the view, not at init: a display link on the main runloop keeps the screen
    // awake for as long as it lives.
    Function("startUiFpsTracking") { self.uiFps.start() }
    Function("stopUiFpsTracking") { self.uiFps.stop() }
    /** -1 until the first window closes, so "not measured" stays distinct from a real zero. */
    Function("getUiFps") { self.uiFps.fps }

    /**
     Epoch milliseconds, so JS can line these up with its own `Date.now()` readings. Returning a map
     rather than constants keeps it a single call and lets either value be absent.
     */
    Function("getStartupTimestamps") { () -> [String: Double?] in
      [
        "processStartMs": processStartEpochMs(),
        "nativeModuleInitMs": self.moduleInitEpochMs,
      ]
    }

    /**
     Busy-waits on the main queue. Sleeping would suspend the thread and let the OS schedule other
     work, which is not what a blocked UI thread looks like — spinning is the accurate simulation.
     Dispatched async so the call returns immediately and JS stays responsive while the UI freezes,
     which is the asymmetry being demonstrated.
     */
    /**
     iOS deliberately exposes less than Android here. There is no system-wide "free RAM" — the closest
     honest number is `os_proc_available_memory()`, how much this process may still allocate before the
     system kills it, which is the figure that actually matters to an app.
     */
    Function("getMemoryMetrics") { () -> [String: Double?] in
      [
        "appBytes": appMemoryFootprintBytes(),
        "totalBytes": Double(ProcessInfo.processInfo.physicalMemory),
        "availableToAppBytes": Double(os_proc_available_memory()),
      ]
    }

    /**
     There is deliberately no `getStorageMetrics` here.

     `FileManager.attributesOfFileSystem` (`.systemSize`, `.systemFreeSize`) is one of Apple's
     required-reason APIs. It needs no runtime permission and shows no prompt, but it obliges a
     `PrivacyInfo.xcprivacy` declaration at submission — and shipping it in a library risks pushing that
     obligation onto every app that embeds this package, for a number that is not about performance.

     Android has no equivalent restriction, so storage is reported there and simply absent here.
     */

    Function("installCrashHandler") {
      // Read the incumbent first: whoever installed before us keeps working, and installing twice
      // would otherwise make us our own "previous" handler and loop.
      if previousExceptionHandler == nil {
        previousExceptionHandler = NSGetUncaughtExceptionHandler()
        NSSetUncaughtExceptionHandler(axonpackExceptionHandler)
      }
    }

    /// Destructive by design: a record still in the file outlived the process that wrote it, which is
    /// the only evidence we have that the app actually died rather than carried on.
    Function("drainPendingCrashes") { () -> [String] in
      guard let url = crashFileURL(), let contents = try? String(contentsOf: url, encoding: .utf8)
      else {
        return []
      }
      try? FileManager.default.removeItem(at: url)
      return contents.split(separator: "\n").map(String.init).filter { !$0.isEmpty }
    }

    /// Used for JS fatals, which RN may follow with the process going down before anyone reads them.
    Function("persistCrashRecord") { (json: String) in
      appendCrashLine(json)
    }

    Function("getDeviceInfo") { () -> [String: Any?] in
      let bundle = Bundle.main.infoDictionary ?? [:]
      #if targetEnvironment(simulator)
        let simulator = true
      #else
        let simulator = false
      #endif
      return [
        "platform": "ios",
        "osVersion": UIDevice.current.systemVersion,
        "model": deviceModelIdentifier(),
        "brand": "Apple",
        "appVersion": bundle["CFBundleShortVersionString"] as? String,
        "buildVersion": bundle["CFBundleVersion"] as? String,
        "bundleId": Bundle.main.bundleIdentifier,
        "isEmulator": simulator,
        "totalMemoryBytes": Double(ProcessInfo.processInfo.physicalMemory),
        "availableMemoryBytes": Double(os_proc_available_memory()),
      ]
    }

    Function("blockMainThread") { (durationMs: Double) in
      DispatchQueue.main.async {
        let deadline = Date().addingTimeInterval(durationMs / 1000)
        while Date() < deadline {
        }
      }
    }

    Function("crashMainThread") { (message: String) in
      DispatchQueue.main.async {
        NSException(          name: NSExceptionName("AxonpackDevtoolsCrash"),
          reason: message,
          userInfo: nil
        ).raise()
      }
    }
  }
}

File: packages/@axonpack/expo-devtools/android/src/main/java/expo/modules/axonpackdevtools/AxonpackDevtoolsModule.kt
package expo.modules.axonpackdevtools

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import android.os.Debug
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.os.Process
import android.os.StatFs
import android.os.SystemClock
import android.view.Choreographer
import com.facebook.react.modules.network.OkHttpClientProvider
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.IOException
import java.io.PrintWriter
import java.io.StringWriter
import okhttp3.Call
import okhttp3.Connection
import okhttp3.EventListener
import org.json.JSONArray
import org.json.JSONObject

/**
 * The only native code in this package, and it exists to *cause* problems rather than measure them:
 * the Limiter section needs to stall or kill the main thread, which JavaScript cannot reach.
 *
 * Optional on the JS side, so an app without a dev client keeps working — it just can't reach the
 * main thread.
 */
/**
 * Frames the *main* thread actually rendered. A `requestAnimationFrame` loop in JS cannot see this: the
 * UI thread can be fully stalled while JS keeps ticking, which is the blind spot the FPS card warns
 * about. `Choreographer` posts on the main looper, so it stops counting exactly when the UI stops
 * drawing.
 */
private class UiFpsTracker : Choreographer.FrameCallback {
  private var frames = 0
  private var windowStartNanos = 0L
  private var running = false

  @Volatile
  var fps: Double = -1.0
    private set

  /** Must be posted to the main looper — `Choreographer.getInstance()` is per-thread. */
  fun start() {
    Handler(Looper.getMainLooper()).post {
      if (running) return@post
      running = true
      frames = 0
      windowStartNanos = System.nanoTime()
      Choreographer.getInstance().postFrameCallback(this)
    }
  }

  fun stop() {
    Handler(Looper.getMainLooper()).post {
      running = false
      fps = -1.0
    }
  }

  override fun doFrame(frameTimeNanos: Long) {
    if (!running) return
    frames += 1
    val elapsedSeconds = (frameTimeNanos - windowStartNanos) / 1_000_000_000.0
    if (elapsedSeconds >= 0.5) {
      fps = frames / elapsedSeconds
      frames = 0
      windowStartNanos = frameTimeNanos
    }
    Choreographer.getInstance().postFrameCallback(this)
  }
}


/**
 * Crash persistence.
 *
 * An uncaught Java or Kotlin exception takes the process down before JavaScript gets another turn, so
 * a report has to be written from the dying thread — the one part of crash reporting JS cannot do for
 * itself. Written as JSON Lines and appended: reading a file back in to rewrite it is the wrong thing
 * to be doing in a crash handler.
 *
 * `Thread.setDefaultUncaughtExceptionHandler` runs in ordinary context, so normal file IO is safe
 * here. NDK/native signal crashes are deliberately out of scope; catching those needs an
 * async-signal-safe handler that would also fight Crashlytics and Sentry over the same slot.
 */
private const val CRASH_FILE_NAME = "axonpack-devtools-crashes.jsonl"

private class CrashPersistingHandler(
  private val file: File,
  /** Whatever was installed before us — React Native's own, or a crash SDK's. */
  private val previous: Thread.UncaughtExceptionHandler?,
) : Thread.UncaughtExceptionHandler {

  override fun uncaughtException(thread: Thread, throwable: Throwable) {
    try {
      file.appendText(describe(thread, throwable) + "\n")
    } catch (_: Throwable) {
      // A failed report must not change how the app dies.
    }
    // Always delegate: swallowing this would leave the process alive in an undefined state, and
    // would take the app's real crash reporter down with it.
    previous?.uncaughtException(thread, throwable)
  }

  private fun describe(thread: Thread, throwable: Throwable): String {
    val trace = StringWriter()
    throwable.printStackTrace(PrintWriter(trace))

    val frames = JSONArray()
    for (element in throwable.stackTrace) frames.put(element.toString())

    return JSONObject()
      .put("kind", "native-exception")
      .put("name", throwable.javaClass.name)
      .put("message", throwable.message ?: "")
      .put("stack", trace.toString())
      .put("timestamp", System.currentTimeMillis())
      .put(
        "native",
        JSONObject()
          .put("type", throwable.javaClass.name)
          .put("thread", thread.name)
          .put("frames", frames),
      )
      .toString()
  }
}


/**
 * Connection-phase timing.
 *
 * The phases inside a request — queueing, DNS, TCP, TLS — are measured by OkHttp and reported nowhere
 * JavaScript can reach. React Native's own `PerformanceResourceTiming` looks like the JS answer and is
 * not: it fills those fields from the same three instants a JS patch already sees, so queueing comes
 * out as zero and connection setup as the whole wait. OkHttp's `EventListener` is the real
 * measurement, and it is a documented extension point rather than a patch — one listener per call,
 * told when each phase begins and ends.
 *
 * Durations come from `nanoTime`, which is monotonic, while the one wall-clock reading is taken at the
 * start so JavaScript can line the record up with the row it belongs to.
 */
private class PhaseEventListener(private val emit: (Map<String, Any?>) -> Unit) : EventListener() {
  private var startedAtEpochMs = 0.0
  private var callStart = 0L
  private var dnsStart = 0L
  private var dnsEnd = 0L
  private var connectStart = 0L
  private var secureStart = 0L
  private var secureEnd = 0L
  private var connectEnd = 0L
  private var requestStart = 0L
  private var requestEnd = 0L
  private var responseStart = 0L
  private var responseEnd = 0L
  private var reusedConnection = true
  private var protocol: String? = null
  private var url: String? = null
  private var decodedBytes = -1L

  /** Null rather than zero for a phase that never happened — a reused connection has no handshake. */
  private fun millis(from: Long, to: Long): Double? =
      if (from == 0L || to == 0L || to < from) null else (to - from) / 1_000_000.0

  override fun callStart(call: Call) {
    startedAtEpochMs = System.currentTimeMillis().toDouble()
    callStart = System.nanoTime()
    url = call.request().url.toString()
  }

  override fun dnsStart(call: Call, domainName: String) {
    dnsStart = System.nanoTime()
  }

  override fun dnsEnd(call: Call, domainName: String, inetAddressList: List<java.net.InetAddress>) {
    dnsEnd = System.nanoTime()
  }

  override fun connectStart(
      call: Call,
      inetSocketAddress: java.net.InetSocketAddress,
      proxy: java.net.Proxy
  ) {
    // Reaching here at all means a connection had to be opened for this call.
    reusedConnection = false
    connectStart = System.nanoTime()
  }

  override fun secureConnectStart(call: Call) {
    secureStart = System.nanoTime()
  }

  override fun secureConnectEnd(call: Call, handshake: okhttp3.Handshake?) {
    secureEnd = System.nanoTime()
  }

  override fun connectEnd(
      call: Call,
      inetSocketAddress: java.net.InetSocketAddress,
      proxy: java.net.Proxy,
      protocol: okhttp3.Protocol?
  ) {
    connectEnd = System.nanoTime()
    this.protocol = protocol?.toString()
  }

  override fun connectionAcquired(call: Call, connection: Connection) {
    if (protocol == null) protocol = connection.protocol().toString()
  }

  override fun requestHeadersStart(call: Call) {
    requestStart = System.nanoTime()
  }

  override fun requestHeadersEnd(call: Call, request: okhttp3.Request) {
    requestEnd = System.nanoTime()
  }

  override fun requestBodyEnd(call: Call, byteCount: Long) {
    // A request with a body is not sent until the body is, so this is the later of the two.
    requestEnd = System.nanoTime()
  }

  override fun responseHeadersStart(call: Call) {
    responseStart = System.nanoTime()
  }

  override fun responseBodyEnd(call: Call, byteCount: Long) {
    responseEnd = System.nanoTime()
    // What the caller read, which is the decoded body: OkHttp gunzips transparently below this.
    decodedBytes = byteCount
  }

  override fun callEnd(call: Call) {
    report()
  }

  override fun callFailed(call: Call, ioe: IOException) {
    // Still reported: how far a request got before it failed is the most useful thing about it.
    report()
  }

  private fun report() {
    val target = url ?: return
    // Whichever phase came first is where the call stopped waiting to be worked on.
    val workStart = listOf(dnsStart, connectStart, requestStart).firstOrNull { it != 0L }
    emit(
        mapOf(
            "url" to target,
            "startMs" to startedAtEpochMs,
            "queuedMs" to millis(callStart, workStart ?: 0L),
            "dnsMs" to millis(dnsStart, dnsEnd),
            // TLS is measured on its own, so the connect phase stops where the handshake starts.
            "tcpMs" to millis(connectStart, if (secureStart != 0L) secureStart else connectEnd),
            "tlsMs" to millis(secureStart, secureEnd),
            "sendMs" to millis(requestStart, requestEnd),
            "waitMs" to millis(if (requestEnd != 0L) requestEnd else requestStart, responseStart),
            "downloadMs" to millis(responseStart, responseEnd),
            "reusedConnection" to reusedConnection,
            "protocol" to protocol,
            // The wire count comes from the interceptor below rather than from here, because by the
            // time a body reaches this listener OkHttp has already decoded it.
            "wireBytes" to WireSizes.take(target),
            "decodedBytes" to if (decodedBytes >= 0) decodedBytes.toDouble() else null,
            "measuredBy" to "okhttp",
        ))
  }
}

/**
 * How many bytes a response actually carried, before OkHttp decoded it.
 *
 * A *network* interceptor sits below transparent gzip and so sees the encoded body and the
 * `Content-Length` that describes it; an application interceptor — and the event listener above —
 * sees what the caller reads, already decoded. Both numbers are needed and only one is visible from
 * each side, so the interceptor leaves its reading here for the listener to collect.
 *
 * Keyed by URL and taken once, because that is all the two sides share. A reading nobody collects is
 * dropped when the map is at its cap rather than growing without bound.
 */
private object WireSizes {
  private const val MAX_PENDING = 64
  private val pending = java.util.concurrent.ConcurrentHashMap<String, Long>()

  fun put(url: String, bytes: Long) {
    if (pending.size >= MAX_PENDING) pending.clear()
    pending[url] = bytes
  }

  fun take(url: String): Double? = pending.remove(url)?.toDouble()
}

/** Whether the listener has been put in front of React Native's client factory already. */
private var networkTimingInstalled = false

class AxonpackDevtoolsModule : Module() {
  /** Captured when the module is constructed, which happens during native startup. */
  private val moduleInitEpochMs = System.currentTimeMillis().toDouble()

  /**
   * `getStartUptimeMillis` is on the uptime clock, so it's shifted onto the epoch to be comparable with
   * `Date.now()` in JS. Requires API 24, which is this module's `minSdkVersion`.
   */
  private fun processStartEpochMs(): Double =
    (System.currentTimeMillis() - SystemClock.uptimeMillis() + Process.getStartUptimeMillis())
      .toDouble()

  private val uiFps = UiFpsTracker()

  private var crashHandlerInstalled = false

  private fun crashFile(): File? =
    appContext.reactContext?.filesDir?.let { File(it, CRASH_FILE_NAME) }

  override fun definition() = ModuleDefinition {
    Name("AxonpackDevtools")

    // Started and stopped by the view, not at init: a frame callback keeps the main thread busy for as
    // long as it lives.
    Events("onNetworkPhases")

    /**
     * Returns whether phases will actually be reported, which is what the Timing tab says out loud.
     *
     * The listener is installed by replacing React Native's OkHttp client *factory*, which is the
     * documented seam — but `NetworkingModule` asks that factory for a client once, when JavaScript
     * first touches networking. So this has to run before the app's first request, which is why the
     * client installs it from `init()` and not when the panel opens. A request made before then is a
     * row without phases rather than a wrong one.
     */
    Function("installNetworkTimingReporter") { ->
      if (networkTimingInstalled) return@Function true
      try {
        OkHttpClientProvider.setOkHttpClientFactory {
          OkHttpClientProvider.createClientBuilder()
              // Below transparent gzip, which is the only place the encoded size is still visible.
              .addNetworkInterceptor { chain ->
                val response = chain.proceed(chain.request())
                val wire =
                    response.header("content-length")?.toLongOrNull()
                        ?: response.body?.contentLength()?.takeIf { it >= 0 }
                if (wire != null) WireSizes.put(response.request.url.toString(), wire)
                response
              }
              .eventListenerFactory(
                  object : EventListener.Factory {
                    override fun create(call: Call): EventListener =
                        PhaseEventListener { payload -> sendEvent("onNetworkPhases", payload) }
                  })
              .build()
        }
        networkTimingInstalled = true
        true
      } catch (error: Throwable) {
        // An OkHttp or React Native this build does not match is a missing tier, not a failure.
        false
      }
    }

    Function("startUiFpsTracking") { uiFps.start() }
    Function("stopUiFpsTracking") { uiFps.stop() }
    /** -1 until the first window closes, so "not measured" stays distinct from a real zero. */
    Function("getUiFps") { uiFps.fps }

    /** Epoch milliseconds, so JS can line these up with its own `Date.now()` readings. */
    Function("getStartupTimestamps") {
      mapOf(
        "processStartMs" to processStartEpochMs(),
        "nativeModuleInitMs" to moduleInitEpochMs,
      )
    }

    /**
     * Busy-waits on the main looper. `Thread.sleep` would suspend and let the OS schedule other work,
     * which is not what a blocked UI thread looks like — spinning is the accurate simulation. Posted
     * rather than run inline so the call returns at once and JS stays responsive while the UI freezes.
     */
    /**
     * `totalPss` is the app's proportional set size in kilobytes — the share of physical memory
     * attributable to this process, which is what a user means by "memory" and is unrelated to the JS
     * heap read from Hermes. Android, unlike iOS, also reports system-wide free RAM.
     */
    Function("getMemoryMetrics") {
      val activityManager =
        appContext.reactContext?.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
      val info = ActivityManager.MemoryInfo()
      activityManager?.getMemoryInfo(info)

      val debugInfo = Debug.MemoryInfo()
      Debug.getMemoryInfo(debugInfo)

      mapOf(
        "appBytes" to debugInfo.totalPss.toDouble() * 1024,
        "totalBytes" to if (activityManager != null) info.totalMem.toDouble() else null,
        "availableToAppBytes" to if (activityManager != null) info.availMem.toDouble() else null,
      )
    }

    /**
     * `StatFs` on the data directory needs no permission and no manifest entry — READ/WRITE_EXTERNAL_
     * STORAGE only ever applied to shared storage. Read on demand rather than sampled: free space moves
     * slowly and crossing the bridge is not free.
     *
     * There is no iOS counterpart on purpose; see the Swift module for why.
     */
    Function("getStorageMetrics") {
      val stat = StatFs(Environment.getDataDirectory().path)
      mapOf(
        "totalBytes" to stat.totalBytes.toDouble(),
        "freeBytes" to stat.availableBytes.toDouble(),
      )
    }

    /** Null until the app's files directory is reachable, which it always is once a context exists. */
    Function("installCrashHandler") {
      val file = crashFile()
      if (file != null && !crashHandlerInstalled) {
        crashHandlerInstalled = true
        Thread.setDefaultUncaughtExceptionHandler(
          CrashPersistingHandler(file, Thread.getDefaultUncaughtExceptionHandler())
        )
      }
    }

    /**
     * Destructive by design: a record still in the file outlived the process that wrote it, which is
     * the only evidence we have that the app actually died rather than carried on.
     */
    Function("drainPendingCrashes") { ->
      val file = crashFile()
      if (file == null || !file.exists()) {
        emptyList<String>()
      } else {
        val lines = file.readLines().filter { it.isNotBlank() }
        file.delete()
        lines
      }
    }

    /** Used for JS fatals, which RN may follow with the process going down before anyone reads them. */
    Function("persistCrashRecord") { json: String ->
      crashFile()?.appendText(json + "\n")
    }

    Function("getDeviceInfo") { ->
      val context = appContext.reactContext
      val packageInfo =
        context?.packageManager?.getPackageInfo(context.packageName, 0)
      val activityManager =
        context?.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
      val memory = ActivityManager.MemoryInfo()
      activityManager?.getMemoryInfo(memory)

      mapOf(
        "platform" to "android",
        "osVersion" to "${Build.VERSION.RELEASE} (API ${Build.VERSION.SDK_INT})",
        "model" to Build.MODEL,
        "brand" to Build.MANUFACTURER,
        "appVersion" to packageInfo?.versionName,
        "buildVersion" to packageInfo?.let {
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) it.longVersionCode.toString()
          else @Suppress("DEPRECATION") it.versionCode.toString()
        },
        "bundleId" to context?.packageName,
        // A heuristic, not a guarantee — there is no supported emulator query, and a wrong `false`
        // is less misleading in a bug report than omitting the field.
        "isEmulator" to (Build.FINGERPRINT.startsWith("generic") ||
          Build.FINGERPRINT.contains("emulator") ||
          Build.MODEL.contains("sdk_gphone")),
        "totalMemoryBytes" to if (activityManager != null) memory.totalMem.toDouble() else null,
        "availableMemoryBytes" to if (activityManager != null) memory.availMem.toDouble() else null,
      )
    }

    Function("blockMainThread") { durationMs: Double ->
      Handler(Looper.getMainLooper()).post {
        val startedAt = System.currentTimeMillis()
        while (System.currentTimeMillis() - startedAt < durationMs.toLong()) {
          // Deliberately empty: occupy the thread rather than yield it.
        }
      }
    }

    Function("crashMainThread") { message: String ->
      Handler(Looper.getMainLooper()).post { throw RuntimeException(message) }
    }
  }
}
